data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java

```diff
@@ -33,7 +35,16 @@
     private final String applicationNameForTracing;
 
     public ClientImpl(ConnectionStringBuilder csb) throws URISyntaxException {
-        clusterUrl = csb.getClusterUrl();
+        String url = csb.getClusterUrl();
+        URI clusterUri = new URI(url);
+        String host = clusterUri.getHost();
+        String auth = clusterUri.getAuthority().toLowerCase();
+        if (host == null && auth.endsWith(";fed=true")) {
+            url = new URIBuilder().setScheme(clusterUri.getScheme()).setHost(auth.substring(0, clusterUri.getAuthority().indexOf(";fed=true"))).toString();
```

Comment:
Need null checks, though I'd personally instead add validation to the CSB constructor for that which is required.

Also, I like how you did this but just to point out that there's a method URIBuilder.fromUri() and then you could just set the truncated host.

Reply:
You're right, CSB can only be created via helper method, as above.